internal/amqp/consumer.go

```diff
@@ -0,0 +1,131 @@
+	}
+
+	if prefetch > 0 {
+		err := channel.Qos(prefetch, 0, false)
+		if err != nil {
+			return nil, errors.WithMessage(err, "set prefetch with qos")
+		}
+	}
+
+	err = declareExchange(channel, exchangeName)
+	if err != nil {
+		return nil, errors.WithMessage(err, "declare exchange")
+	}
+
+	amqpQueue, err := channel.QueueDeclare(
+		queueName, // name
+		true,      // durable
+		false,     // delete when unused
+		false,     // exclusive
+		false,     // no-wait
+		nil,       // arguments
+	)
+	if err != nil {
+		return nil, errors.WithMessage(err, "declare queue")
+	}
+
+	err = channel.QueueBind(queueName, routingKey, exchangeName, false, nil)
+	if err != nil {
+		return nil, errors.WithMessage(err, "bind to queue")
+	}
+	return &consumer{
+		channel: channel,
+		queue:   &amqpQueue,
+	}, nil
+}
+
+func (c *consumer) Close() error {
+	// TODO: stop consuming messages with c.channel.Cancel() and wait for the
+	// consumer to complete before closing the connection to avoid dropping
+	// inflight messages.
+	err := c.channel.Close()
+	if err != nil {
+		return errors.WithMessage(err, "close amqp channel")
+	}
+	return nil
+}
+
+func (c *consumer) Start(logger *log.Logger, handlers map[string]func([]byte) error) error {
+	msgs, err := c.channel.Consume(
+		c.queue.Name,      // queue
+		"release-manager", // consumer
+		false,             // auto-ack
+		false,             // exclusive
+		false,             // no-local
+		false,             // no-wait
+		nil,               // args
+	)
+	if err != nil {
+		return errors.WithMessage(err, "consume messages")
+	}
+
+	for msg := range msgs {
+		logger := logger.With(
+			"routingKey", msg.RoutingKey,
+			"messageId", msg.MessageId,
+			"eventType", msg.Type,
+			"correlationId", msg.CorrelationId,
+			"headers", fmt.Sprintf("%#v", msg.Headers),
+		)
+		logger.Infof("Received message type=%s from exchange=%s routingKey=%s messageId=%s correlationId=%s timestamp=%s", msg.Type, msg.Exchange, msg.RoutingKey, msg.MessageId, msg.CorrelationId, msg.Timestamp)
+		now := time.Now()
+
+		handler, ok := handlers[msg.Type]
+		if !ok {
+			logger.With("res", map[string]interface{}{
+				"status": "failed",
+				"error":  "unprocessable",
+			}).Errorf("[consumer] [UNPROCESSABLE] Failed to handle message: no handler registered for event type '%s': dropping it", msg.Type)
+			err := msg.Nack(false, false)
+			if err != nil {
+				logger.Errorf("Failed to nack message: %v", err)
+			}
+			continue
+		}
+		err := handler(msg.Body)
+		duration := time.Since(now).Milliseconds()
+		if err != nil {
+			logger.With("res", map[string]interface{}{
+				"status":       "failed",
+				"responseTime": duration,
+				"error":        fmt.Sprintf("%+v", err),
+			}).Errorf("[consumer] [FAILED] Failed to handle message: nacking and requeing: %v", err)
+			// TODO: remove comments to allow for redelivery. This will put events
+			// into the unacknowledged state
+
+			// err := msg.Nack(false, true) if err != nil {
```

Comment:
Should the comments be removed? :)

Reply:
Hmm. Yes. The unacked mechanism is critical when having preferch of one, as an error will block releases. But uncommenting can result in looping failed events leading to high loads.

The right thing to do is to add a dead letter queue on the messages. I think we can do that rather simple, but in another PR

Reply:
Sounds good 👍